WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift

```diff
@@ -15,70 +15,77 @@ class PointOfSaleItemsController: PointOfSaleItemsControllerProtocol {
     private(set) var itemListStatePublisher: any Publisher<ItemListState, Never>
     private var itemListStateSubject: PassthroughSubject<ItemListState, Never> = .init()
     private var allItems: [POSItem] = []
-    private var currentPage: Int = Constants.initialPage
-    private var mightHaveMorePages: Bool = true
+    private let paginationTracker: AsyncPaginationTracker
     private let itemProvider: PointOfSaleItemServiceProtocol
 
     init(itemProvider: PointOfSaleItemServiceProtocol) {
         self.itemProvider = itemProvider
+        self.paginationTracker = .init(pageFirstIndex: Constants.initialPage)
         itemListStatePublisher = itemListStateSubject.eraseToAnyPublisher()
     }
 
     @MainActor
     func loadInitialItems() async {
-        mightHaveMorePages = true
         itemListStateSubject.send(.initialLoading)
-        try? await load(pageNumber: Constants.initialPage)
+        do {
+            try await paginationTracker.syncFirstPage { [weak self] pageNumber in
+                guard let self else { return true }
+                return try await fetchItems(pageNumber: pageNumber)
+            }
+            updateItemListStateAfterLoadAttempt()
+        } catch {
+            itemListStateSubject.send(.error(PointOfSaleErrorState.errorOnLoadingProducts()))
+        }
     }
 
     @MainActor
     func loadNextItems() async {
+        itemListStateSubject.send(.loading(allItems))
```

Comment:
```suggestion
        guard paginationTracker.hasNextPage else {
            return
        }
        itemListStateSubject.send(.loading(allItems))
```

You'll also need to make `hasNextPage` a `private(set)` var as well.

Reply:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/14728/commits/a2e792aff49ce36f2ea69078bd6839dfe33666ad, this should fix the flickering issue when loading the next page from the infinite scroll.